File: apps/ios-starter/ThtwaatStarter/Data/Repositories/AuthRepository.swift
import Foundation

actor AuthRepository {
    private let client: APIClient
    private let sessionStore: any SessionStore

    init(client: APIClient, sessionStore: any SessionStore) {
        self.client = client
        self.sessionStore = sessionStore
    }

    func login(email: String, password: String) async -> APIResult<TokenResponse> {
        let result: APIResult<TokenResponse> = await client.post(
            "/api/v1/auth/login",
            body: LoginRequest(email: email, password: password),
            auth: false
        )
        if case .success(let tokens) = result {
            await sessionStore.saveTokens(access: tokens.accessToken, refresh: tokens.refreshToken)
        }
        return result
    }

    func signup(email: String, password: String, first: String, last: String) async -> APIResult<UserProfile> {
        await client.post(
            "/api/v1/users/",
            body: SignupRequest(email: email, password: password, firstName: first, lastName: last, companyId: nil),
            auth: false
        )
    }

    func forgotPassword(email: String) async -> APIResult<APIClient.EmptyResponse> {
        await client.post("/api/v1/auth/forgot-password", body: ForgotPasswordRequest(email: email), auth: false)
    }

    func me() async -> APIResult<UserProfile> {
        await client.get("/api/v1/auth/me")
    }

    func logout() async -> APIResult<APIClient.EmptyResponse> {
        let refresh = await sessionStore.refreshToken() ?? ""
        let result: APIResult<APIClient.EmptyResponse> = await client.post(
            "/api/v1/auth/logout",
            body: LogoutRequest(refreshToken: refresh)
        )
        await sessionStore.clearSession()
        return result
    }

    @discardableResult
    func refreshIfPossible() async -> Bool {
        guard let refresh = await sessionStore.refreshToken(), !refresh.isEmpty else { return false }
        let result: APIResult<TokenResponse> = await client.post(
            "/api/v1/auth/refresh",
            body: RefreshRequest(refreshToken: refresh),
            auth: false
        )
        if case .success(let tokens) = result {
            await sessionStore.saveTokens(access: tokens.accessToken, refresh: tokens.refreshToken)
            return true
        }
        await sessionStore.clearSession()
        return false
    }
}

File: apps/android-starter/app/src/main/java/com/thtwaat/starter/data/repository/AuthRepository.kt
package com.thtwaat.starter.data.repository

import com.thtwaat.starter.core.datastore.SessionStore
import com.thtwaat.starter.core.network.safeApiCall
import com.thtwaat.starter.core.util.Result
import com.thtwaat.starter.data.remote.api.AuthApi
import com.thtwaat.starter.data.remote.dto.ForgotPasswordRequestDto
import com.thtwaat.starter.data.remote.dto.LoginRequestDto
import com.thtwaat.starter.data.remote.dto.LogoutRequestDto
import com.thtwaat.starter.data.remote.dto.ResetPasswordRequestDto
import com.thtwaat.starter.data.remote.dto.SignupRequestDto
import com.thtwaat.starter.data.remote.dto.TokenResponseDto
import com.thtwaat.starter.data.remote.dto.UserProfileDto
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val api: AuthApi,
    private val sessionStore: SessionStore,
) {
    val isLoggedIn: Flow<Boolean> = sessionStore.accessToken.map { !it.isNullOrBlank() }

    suspend fun login(email: String, password: String): Result<TokenResponseDto> {
        val result = safeApiCall { api.login(LoginRequestDto(email, password)) }
        if (result is Result.Success) {
            sessionStore.saveTokens(result.data.accessToken, result.data.refreshToken)
        }
        return result
    }

    suspend fun signup(
        email: String,
        password: String,
        firstName: String,
        lastName: String,
        companyId: String? = null,
    ): Result<UserProfileDto> = safeApiCall {
        api.signup(SignupRequestDto(email, password, firstName, lastName, companyId))
    }

    suspend fun forgotPassword(email: String): Result<Unit> =
        safeApiCall { api.forgotPassword(ForgotPasswordRequestDto(email)) }

    suspend fun resetPassword(email: String, code: String, newPassword: String): Result<Unit> =
        safeApiCall { api.resetPassword(ResetPasswordRequestDto(email, code, newPassword)) }

    suspend fun me(): Result<UserProfileDto> = safeApiCall { api.me() }

    suspend fun logout(): Result<Unit> {
        val refresh = sessionStore.getRefreshToken()
        val result = if (refresh.isNullOrBlank()) {
            Result.Success(Unit)
        } else {
            safeApiCall { api.logout(LogoutRequestDto(refresh)) }
        }
        sessionStore.clearSession()
        return result
    }
}
